File: 02/2150.cpp
#include <algorithm>
#include <cstdio>
#include <stack>
#include <vector>

class scc {
   public:
    static std::vector<std::vector<int>> get(
        int n, const std::vector<std::vector<int>> &edge) {
        std::vector<std::vector<int>> scc_vec;
        std::vector<int> dfsn;
        std::vector<bool> finish;
        std::stack<int> stack;
        int count = 0;

        dfsn.resize(n);
        finish.resize(n);

        for (int i = 0; i < n; ++i)
            if (dfsn[i] == 0) dfs(edge, i, scc_vec, dfsn, finish, stack, count);

        return scc_vec;
    };

   private:
    static int dfs(const std::vector<std::vector<int>> &edge, int x,
                   std::vector<std::vector<int>> &scc_vec,
                   std::vector<int> &dfsn, std::vector<bool> &finish,
                   std::stack<int> &stack, int &count) {
        dfsn[x] = ++count;
        stack.push(x);

        int top = dfsn[x];
        for (int to : edge[x]) {
            if (dfsn[to] == 0)
                top = std::min(
                    top, dfs(edge, to, scc_vec, dfsn, finish, stack, count));
            else if (!finish[to])
                top = std::min(top, dfsn[to]);
        }

        if (top == dfsn[x]) {
            std::vector<int> now_scc;
            int v = -1;

            while (v != x) {
                v = stack.top();
                stack.pop();
                finish[v] = true;
                now_scc.push_back(v);
            }

            scc_vec.push_back(now_scc);
        }

        return top;
    }
};

int main() {
    int v, e;
    std::vector<std::vector<int>> edge;

    scanf("%d %d", &v, &e);
    edge.resize(v);

    for (int i = 0; i < e; ++i) {
        int a, b;
        scanf("%d %d", &a, &b);
        --a, --b;
        edge[a].push_back(b);
    }

    auto scc_vec = scc::get(v, edge);

    for (int i = 0; i < (int)scc_vec.size(); ++i)
        std::sort(scc_vec[i].begin(), scc_vec[i].end());
    std::sort(scc_vec.begin(), scc_vec.end());

    printf("%d\n", (int)scc_vec.size());
    for (auto &now : scc_vec) {
        for (int val : now) printf("%d ", val + 1);
        printf("-1\n");
    }
    return 0;
}

File: 02/2150.rs
fn main() {
    let edges: Vec<Vec<usize>> = get_input();

    let mut sccs: Vec<Vec<usize>> = scc::get_scc_vec(&edges);
    for scc in sccs.iter_mut() {
        scc.sort();
    }
    sccs.sort_by(|a, b| a[0].cmp(&b[0]));

    println!("{}", sccs.len());
    for scc in sccs.iter() {
        for i in scc.iter() {
            print!("{} ", i + 1);
        }
        println!("-1");
    }
}

mod scc {
    struct St {
        scc_vec: Vec<Vec<usize>>,
        dfsn: Vec<Option<usize>>,
        stack: Vec<usize>,
        count: usize,
    }

    pub fn get_scc_vec(edge: &Vec<Vec<usize>>) -> Vec<Vec<usize>> {
        let n = edge.len();
        let mut st: St = St {
            scc_vec: Vec::new(),
            dfsn: vec![Some(0); n],
            stack: Vec::new(),
            count: 0,
        };

        for i in 0..n {
            if st.dfsn[i] == Some(0) {
                dfs(i, edge, &mut st);
            }
        }

        st.scc_vec
    }

    fn dfs(x: usize, edge: &Vec<Vec<usize>>, st: &mut St) -> usize {
        st.count += 1;
        st.dfsn[x] = Some(st.count);
        st.stack.push(x);

        let mut top: usize = st.count;
        for to in edge[x].iter() {
            if st.dfsn[*to] == Some(0) {
                top = std::cmp::min(top, dfs(*to, edge, st));
            } else if st.dfsn[*to] != None {
                top = std::cmp::min(top, st.dfsn[*to].unwrap());
            }
        }

        if top == st.dfsn[x].unwrap() {
            let mut now_scc: Vec<usize> = Vec::new();

            loop {
                let v = *(st.stack.last().unwrap());
                st.stack.pop();
                st.dfsn[v] = None;
                now_scc.push(v);
                if v == x {
                    break;
                }
            }

            st.scc_vec.push(now_scc);
        }

        top
    }
}

fn get_input() -> Vec<Vec<usize>> {
    let buf = std::io::read_to_string(std::io::stdin()).unwrap();
    let mut iter = buf.split_whitespace().flat_map(str::parse::<usize>);
    let mut get = || iter.next().unwrap();

    let (v, e) = (get(), get());
    let mut edges: Vec<Vec<usize>> = vec![Vec::new(); v];

    for _ in 0..e {
        let (a, b) = (get() - 1, get() - 1);
        edges[a].push(b);
    }

    edges
}
